### PR TITLE
zephyr: Enable integration as a zephyr module

### PR DESCRIPTION
**Pull request by MaureenHelm**
Adds missing cmake file for the pint driver.

Introduces a Kconfig framework to allow applications to conditionally
build drivers and for devices to declare which drivers they support.

Adds module.yml to specify how to integrate the MCUXpresso SDK into the
zephyr build system.

cc: @galak @henrikbrixandersen

### CODE REVIEW

File: zephyr/Kconfig.devices [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	select HAS_MCUX_SDK_DRIVER_ADC16
+	select HAS_MCUX_SDK_DRIVER_CLOCK
+	select HAS_MCUX_SDK_DRIVER_CMP
+	select HAS_MCUX_SDK_DRIVER_CMT
+	select HAS_MCUX_SDK_DRIVER_COMMON
+	select HAS_MCUX_SDK_DRIVER_CRC
+	select HAS_MCUX_SDK_DRIVER_DAC
+	select HAS_MCUX_SDK_DRIVER_DMAMUX
+	select HAS_MCUX_SDK_DRIVER_DSPI
+	select HAS_MCUX_SDK_DRIVER_DSPI_EDMA
```
Comment (hakehuang) [inline zephyr/Kconfig.devices:40]: Zephyr DMA implementation does not compatible with our DSPI EDMA APIS,  there is no need to include DSPI EDMA driver, see https://github.com/hakehuang/zephyr/blob/spi_edma/drivers/spi/spi_mcux_dspi.c. @mmahadevan108 do you use SDK peripheral DMA API driver in your zephyr driver?

File: zephyr/Kconfig.devices [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	select HAS_MCUX_SDK_DRIVER_EDMA
+	select HAS_MCUX_SDK_DRIVER_EWM
+	select HAS_MCUX_SDK_DRIVER_FLASH
+	select HAS_MCUX_SDK_DRIVER_FLEXCAN
+	select HAS_MCUX_SDK_DRIVER_FTM
+	select HAS_MCUX_SDK_DRIVER_GPIO
+	select HAS_MCUX_SDK_DRIVER_I2C
+	select HAS_MCUX_SDK_DRIVER_I2C_EDMA
```
Comment (hakehuang) [inline zephyr/Kconfig.devices:48]: same as above, we do not need this driver

File: zephyr/Kconfig.devices [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	select HAS_MCUX_SDK_DRIVER_PIT
+	select HAS_MCUX_SDK_DRIVER_PORT
+	select HAS_MCUX_SDK_DRIVER_RCM
+	select HAS_MCUX_SDK_DRIVER_RNGA
+	select HAS_MCUX_SDK_DRIVER_RTC
+	select HAS_MCUX_SDK_DRIVER_SIM
+	select HAS_MCUX_SDK_DRIVER_SMC
+	select HAS_MCUX_SDK_DRIVER_SYSMPU
+	select HAS_MCUX_SDK_DRIVER_UART
+	select HAS_MCUX_SDK_DRIVER_UART_EDMA
```
Comment (hakehuang) [inline zephyr/Kconfig.devices:60]: same here

File: zephyr/all_drivers.cmake [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DMAMUX			driver_dmamux)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DSPI			driver_dspi)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DSPI_EDMA			driver_dspi_edma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_EDMA			driver_edma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_EWM			driver_ewm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLASH			driver_flash)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCAN			driver_flexcan)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM			driver_flexcomm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C		driver_flexcomm_i2c)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C_DMA		driver_flexcomm_i2c_dma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:27]: same here

File: zephyr/all_drivers.cmake [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_CLOCK			driver_clock)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_CMP			driver_cmp)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_CMP_1			driver_cmp_1)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_CMT			driver_cmt)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_COMMON			driver_common)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_CRC			driver_crc)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DAC			driver_dac)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DMAMUX			driver_dmamux)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DSPI			driver_dspi)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_DSPI_EDMA			driver_dspi_edma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:20]: same here

File: zephyr/all_drivers.cmake [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_EDMA			driver_edma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_EWM			driver_ewm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLASH			driver_flash)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCAN			driver_flexcan)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM			driver_flexcomm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C		driver_flexcomm_i2c)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C_DMA		driver_flexcomm_i2c_dma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2S		driver_flexcomm_i2s)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2S_DMA		driver_flexcomm_i2s_dma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:29]: same here

File: zephyr/all_drivers.cmake [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_SPI		driver_flexcomm_spi)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_SPI_DMA		driver_flexcomm_spi_dma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_USART		driver_flexcomm_usart)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_USART_DMA		driver_flexcomm_usart_dma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:33]: same here

File: zephyr/all_drivers.cmake [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_EWM			driver_ewm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLASH			driver_flash)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCAN			driver_flexcan)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM			driver_flexcomm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C		driver_flexcomm_i2c)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2C_DMA		driver_flexcomm_i2c_dma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2S		driver_flexcomm_i2s)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_I2S_DMA		driver_flexcomm_i2s_dma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_SPI		driver_flexcomm_spi)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_SPI_DMA		driver_flexcomm_spi_dma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:31]: same here

File: zephyr/all_drivers.cmake [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_USART		driver_flexcomm_usart)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FLEXCOMM_USART_DMA		driver_flexcomm_usart_dma)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_FTM			driver_ftm)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_GPIO			driver_gpio)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_I2C			driver_i2c)
+include_ifdef(CONFIG_MCUX_SDK_DRIVER_I2C_EDMA			driver_i2c_edma)
```
Comment (hakehuang) [inline zephyr/all_drivers.cmake:37]: same here

**Review by hakehuang — APPROVED** [review]
In Zephyr the DMA architecture is not compatible with SDK peripheral EDMA/DMA driver, so we actually do not need those special drivers. but there is no harm to still configure them here, just for your information.

**Comment by MaureenHelm** [comment]
> In Zephyr the DMA architecture is not compatible with SDK peripheral EDMA/DMA driver, so we actually do not need those special drivers. but there is no harm to still configure them here, just for your information.

My goal is to cover all of the drivers in the mcux-sdk repo, whether we currently use them in Zephyr or not, because I think other non-Zephyr applications could use them. I started with drivers supported on K64F and LPC55S69 to demonstrate the concept, and will add the others for i.MX etc. if we agree this concept makes sense.

**Review by dleach02 — APPROVED** [review]

**Comment by nxp-susan** [comment]
> > In Zephyr the DMA architecture is not compatible with SDK peripheral EDMA/DMA driver, so we actually do not need those special drivers. but there is no harm to still configure them here, just for your information.
> 
> My goal is to cover all of the drivers in the mcux-sdk repo, whether we currently use them in Zephyr or not, because I think other non-Zephyr applications could use them. I started with drivers supported on K64F and LPC55S69 to demonstrate the concept, and will add the others for i.MX etc. if we agree this concept makes sense.

Thanks for the introducing such concept. I see you would like to directly use mcux-sdk for zephyr support instead of current hal-nxp module, am I right?  I was little afraid if this will make non -zephyr user feel confused but it could still OK for current prototype you are proposing in the pr: the zephyr needed files are in separate folder zephyr/ under mcux-sdk, the code size is acceptable and SDK regular release delivery will not touch these files.
I see you have added Kconfig files for zephyr, recently we have internal discussion on whether to do Kconfig support for mcux-sdk, because this fills the gap on how user select driver/components he wants for using a SoC/board.  Currently users are expected to learn the all_devices.cmake and all_lib_device_$SOC.cmake to find the information. Having Kconfig could greatly help user fur such situation. What do you see if we have Kconfig support in mcux-sdk? And if we have Kconfig support, zephyr support may directly reference our Kconfig files instead of prepare its own.

**Comment by MaureenHelm** [comment]
> Thanks for the introducing such concept. I see you would like to directly use mcux-sdk for zephyr support instead of current hal-nxp module, am I right?

Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.

> What do you see if we have Kconfig support in mcux-sdk?

I don't quite understand your question?

**Comment by henrikbrixandersen** [comment]
> Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.

Are there plans for supporting all the devices supported by hal-nxp in mcux-sdk going forward or will we be stuck with two NXP HALs forever?

**Comment by nxp-susan** [comment]
> > Thanks for the introducing such concept. I see you would like to directly use mcux-sdk for zephyr support instead of current hal-nxp module, am I right?
> 
> Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.
> 
> > What do you see if we have Kconfig support in mcux-sdk?
> 
> I don't quite understand your question?

I want to know if mcux-sdk introduce Kconfig support, will this help project like zephyr for the configuration requirement to pick up software components? And is there specific requirement for the Kconfig from the point of zephyr project?

**Comment by nxp-susan** [comment]
> > Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.
> 
> Are there plans for supporting all the devices supported by hal-nxp in mcux-sdk going forward or will we be stuck with two NXP HALs forever?

I guess the answer would be to use mcux-sdk for all devices support in future, @MaureenHelm to comment.

**Comment by MaureenHelm** [comment]
> I want to know if mcux-sdk introduce Kconfig support, will this help project like zephyr for the configuration requirement to pick up software components?

Yes, that is the motivation for this PR.

> And is there specific requirement for the Kconfig from the point of zephyr project?

The Kconfig symbols can't conflict with any symbols defined within Zephyr, which is why I prefixed everything in this PR with `MCUX_SDK_`

> > > Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.
> > 
> > 
> > Are there plans for supporting all the devices supported by hal-nxp in mcux-sdk going forward or will we be stuck with two NXP HALs forever?
> 
> I guess the answer would be to use mcux-sdk for all devices support in future, @MaureenHelm to comment.

This would be ideal, yes. @henrikbrixandersen is interested in KE18F which was recently dropped in SDK 2.9 and therefore isn't included in this repo.

**Comment by henrikbrixandersen** [comment]
> This would be ideal, yes. @henrikbrixandersen is interested in KE18F which was recently dropped in SDK 2.9 and therefore isn't included in this repo.

Well, we're interested in the entire KE1xF family. We have quite a few products based on these, all running Zephyr.

**Comment by nxp-susan** [comment]
> > I want to know if mcux-sdk introduce Kconfig support, will this help project like zephyr for the configuration requirement to pick up software components?
> 
> Yes, that is the motivation for this PR.
> 
> > And is there specific requirement for the Kconfig from the point of zephyr project?
> 
> The Kconfig symbols can't conflict with any symbols defined within Zephyr, which is why I prefixed everything in this PR with `MCUX_SDK_`
> 
> > > > Yes, that's right. We may need to keep the hal-nxp module for older devices not supported in mcux-sdk.
> > > 
> > > 
> > > Are there plans for supporting all the devices supported by hal-nxp in mcux-sdk going forward or will we be stuck with two NXP HALs forever?
> > 
> > 
> > I guess the answer would be to use mcux-sdk for all devices support in future, @MaureenHelm to comment.
> 
> This would be ideal, yes. @henrikbrixandersen is interested in KE18F which was recently dropped in SDK 2.9 and therefore isn't included in this repo.

Thanks Maureen. Is that OK we hold the PR for now? Because after mcux-sdk has added Kconfig feature, updates are needed in this PR to leverage the mcux-sdk Kconfig file.
And for KE18F in @henrikbrixandersen's case, we are aware of the device support gap between zephyr scope and mcux release scope, and we plan to fill this gap by enable the devices directly in mcux-sdk main branch. But I cannot commit the schedule for now. I have already created the task for KE18F, appreciate if you could provide a device list(by priority) you would like mcux-sdk to add to support zephyr.

**Pull request closed without merging**